core/src/main/scala/kafka/server/AbstractFetcherManager.scala

```diff
@@ -188,49 +193,57 @@ abstract class AbstractFetcherManager[T <: AbstractFetcherThread](val name: Stri
           tp -> OffsetAndEpoch(brokerAndInitOffset.initOffset, brokerAndInitOffset.currentLeaderEpoch)
         }
 
-        addPartitionsToFetcherThread(fetcherThread, initialOffsetAndEpochs)
+        addPartitionFutures += addPartitionsToFetcherThread(fetcherThread, initialOffsetAndEpochs)
       }
+
+      // wait for all futures to finish
+      KafkaFuture.allOf(addPartitionFutures:_*).get()
     }
   }
 
   protected def addPartitionsToFetcherThread(fetcherThread: T,
-                                             initialOffsetAndEpochs: collection.Map[TopicPartition, OffsetAndEpoch]): Unit = {
-    fetcherThread.addPartitions(initialOffsetAndEpochs)
+                                             initialOffsetAndEpochs: collection.Map[TopicPartition, OffsetAndEpoch]): KafkaFuture[Void] = {
+    val future = fetcherThread.addPartitions(initialOffsetAndEpochs)
     info(s"Added fetcher to broker ${fetcherThread.sourceBroker.id} for partitions $initialOffsetAndEpochs")
+    future
   }
 
   def removeFetcherForPartitions(partitions: Set[TopicPartition]): Unit = {
     lock synchronized {
+      val futures = mutable.Buffer[KafkaFuture[Void]]()
       for (fetcher <- fetcherThreadMap.values)
-        fetcher.removePartitions(partitions)
+        futures += fetcher.removePartitions(partitions)
+      KafkaFuture.allOf(futures:_*).get()
       failedPartitions.removeAll(partitions)
     }
+
     if (partitions.nonEmpty)
       info(s"Removed fetcher for partitions $partitions")
   }
 
   def shutdownIdleFetcherThreads(): Unit = {
     lock synchronized {
-      val keysToBeRemoved = new mutable.HashSet[BrokerIdAndFetcherId]
+      val futures = mutable.Map[BrokerIdAndFetcherId, KafkaFuture[Int]]()
+
       for ((key, fetcher) <- fetcherThreadMap) {
-        if (fetcher.idle) {
-          fetcher.shutdown()
-          keysToBeRemoved += key
+        futures.put(key, fetcher.getPartitionsCount())
```

Comment:
this function is called on every leaderAndIsr request as part of "becomeLeaderOrFollower". So how much delay can be introduced by sending "getPartitionsCount" request to event manager and waiting for result?

Reply:
In upstream Kafka, the request handler thread needs to compete for the partitionMapLock in order to get the partitions count. And the fetcher.idle variable is a hotfix made internally by us. 
When switching to the event-based model, I feel it's cleaner to request the partition count using an event + future.
I'm not sure about the performance difference yet, but I have plans to eventually merge the remove partition, add partition and get partition count into one event if necessary. So ultimately the extra delay introduced by the GetPartitionCount event may not matter.